prototype-spec.md

```diff
@@ -0,0 +1,253 @@
+
+## Root metadata
+
+Root metadata should be signed using an offline process. It can then be uploaded to the registry by running `nv2 upload root`.
+
+## Signing a manifest
+
+To upload a manifest signed with Notary v2, a developer must:
+* Generate and sign targets metadata by running `nv2 sign imagename keyloc --role targetsrolename` where `imagename` is the manifest they are signing, `targetsrolename` is the name of the top-level or delegated targets role the developer is a part of, and `keyloc` is a path to the developer's private key.
+* Upload the manifest and targets metadata to the registry by running `nv2 upload imagename`.
+* Send the targets metadata to the snapshot process. This may be part of the above command, but the developer should specify a `SNAPSHOT_LOCATION`.
+
+## Configuring delegations
+
+As developers join or leave teams, it may be necessary to update delegations. To do so, the delegator should perform the following steps:
+* To add a new delegatee they run `nv2 delegate fromrole torole --namespace ns --key keyloc`. This should add `torole` as a delegation to `fromrole` that is trusted for images in the namespace `ns`.
+* To remove a delegatee they run `nv2 revoke fromrole torole --key keyloc`.
+* After performing the above commands any number of times, they run `nv2 upload fromrole`. This will upload the new `fromrole` metadata. Any new delegated roles should sign and upload their targets metadata as well.
+
+
+## Snapshot and timestamp process
+
+For each TUF root, there must be a process that generates snapshot and timestamp metadata. This process should be automated and may be hosted by a registry or an organization that uploads to a registry. It must include the following steps:
+
+* Receive new metadata version. A developer who plans to upload new TUF targets metadata must send information about the version number of this new metadata to the snapshot process. The process may perform any verification to ensure the developers are authorized to upload a new version.
+* Add metadata version to snapshot. The version number of the new metadata must be added to snapshot and associated with the correct metadata file.
+* Sign snapshot metadata. This may be done using an online key.
+* Update timestamp metadata with the new hash of the snapshot metadata.
+* Sign timestamp metadata. This may be done using an online key.
+* Upload snapshot and timestamp metadata to the registry.
+
+Changes to snapshot may be batched, so multiple metadata versions may be added at once. To do so, the process could allow steps 1 and 2 to repeat any number of times before signing the snapshot metadata.
+## User download
+
+When installing an image from a registry, a user may run `nv2 download imagename`. This command must download the following from the registry:
+* Root metadata. The client should download the most recent root metadata, as well as previous versions until their currently trusted root version.
```

Comment:
Root metadata. The client should download the most recent root metadata, as well as all previous versions until they reach their currently trusted root version.

Reply:
> Does the root metadata span multiple repositories? Assuming it does, we'll need to figure out how a client pulling from one repository knows where to look for the root metadata if that's in another repository. Or we'll need to copy the root metadata to each repository that has signed images.

That's a good question. To allow artifacts to be copied between repositories and registries, it would be nice to have support for a cross-registry root.  The client could have some configuration info to point to the registry with the root metadata. Another option would be to have different roots for each registry, and require that targets metadata on the copied-to registry delegate to any copied artifacts (which could help ensure the copy was validated).

Does either of these sound simpler to use and maintain?

Reply:
> That's a good question. To allow artifacts to be copied between repositories and registries, it would be nice to have support for a cross-registry root. The client could have some configuration info to point to the registry with the root metadata. Another option would be to have different roots for each registry, and require that targets metadata on the copied-to registry delegate to any copied artifacts (which could help ensure the copy was validated).

For now I'm focused on just the repository without expanding to multiple registries, but the extra step is a minor implementation detail. The challenge I'm looking to resolve is that all registry API calls are scoped to a repository. There's no concept of querying a registry for an artifact, you have to specify the repository in the request.

So if the root metadata is not in the repository with the images (where the snapshot and timestamp metadata would be in the same repository as the image being signed and verified), we're going to need some kind of pointer to tell the client where to go looking for the root metadata. I'd rather that be an artifact in the registry that says "my root metadata is in this repository" rather than a per-client configuration.

The other option is copying the root metadata to each repository, which could be a bit of a hassle when updating the root metadata, but I don't know TUF well enough to know how often that would happen. On the registry server, all these copies of the root metadata would be a no-op, just a server side blob mount that creates another pointer to the same sha. So it's more a question of whether we want to trigger lots of API calls when the root metadata could cover a large number of repositories to update.

Reply:
Root metadata shouldn't have to be updated very frequently, but I like the idea of a placeholder/pointer in each repository that points to where the metadata is stored. Maybe the registry could have a 'tuf' repository that hosts the top-level metadata and can be referenced by any other repositories that use Notary signatures.

Reply:
I added this idea to the spec. We can iterate as needed.